objects/src/assets/mod.rs

```diff
@@ -0,0 +1,420 @@
+    pub const MAX_AMOUNT: u64 = (1u64 << 63) - 1;
+
+    // CONSTRUCTOR
+    // --------------------------------------------------------------------------------------------
+    /// Returns a fungible asset instantiated with the provided faucet ID and amount.
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// - The provided faucet ID is not for a fungible asset faucet.
+    /// - The provide amount is greater than 2^63 - 1.
+    pub fn new(faucet_id: AccountId, amount: u64) -> Result<Self, AssetError> {
+        if !faucet_id.is_fungible_faucet() {
+            return Err(AssetError::not_a_fungible_faucet_id(faucet_id));
+        }
+
+        if amount > Self::MAX_AMOUNT {
+            return Err(AssetError::amount_too_big(amount));
+        }
+
+        Ok(Self { faucet_id, amount })
+    }
+
+    // PUBLIC ACCESSORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Return ID of the faucet which issued this asset.
+    pub fn faucet_id(&self) -> AccountId {
+        self.faucet_id
+    }
+
+    /// Returns the amount of this asset.
+    pub fn amount(&self) -> u64 {
+        self.amount
+    }
+
+    // OPERATIONS
+    // --------------------------------------------------------------------------------------------
+
+    /// Adds two fungible assets together and returns the result.
+    ///
+    /// # Errors
+    /// Returns an error if the assets were not issued by the same faucet.
+    #[allow(clippy::should_implement_trait)]
+    pub fn add(self, other: Self) -> Result<Self, AssetError> {
+        if self.faucet_id != other.faucet_id {}
+
+        let amount = self.amount + other.amount;
+        if amount > Self::MAX_AMOUNT {
+            return Err(AssetError::amount_too_big(amount));
+        }
+
+        Ok(Self {
+            faucet_id: self.faucet_id,
+            amount,
+        })
+    }
+
+    /// Subtracts the specified amount from this asset and returns the resulting asset.
+    ///
+    /// # Errors
+    /// Returns an error if this asset's amount is smaller than the requested amount.
+    pub fn sub(&mut self, amount: u64) -> Result<Self, AssetError> {
+        if amount > self.amount {
+            return Err(AssetError::AmountNotSufficient(self.amount, amount));
+        }
+
+        self.amount -= amount;
+        Ok(FungibleAsset {
+            faucet_id: self.faucet_id,
+            amount,
+        })
+    }
+}
+
+impl From<FungibleAsset> for Word {
+    fn from(asset: FungibleAsset) -> Self {
+        let mut result: Word = asset.faucet_id.into();
+        debug_assert_eq!(result[3], ZERO);
+        result[3] = Felt::new(asset.amount);
+        result
+    }
+}
+
+impl From<FungibleAsset> for [u8; 32] {
+    fn from(asset: FungibleAsset) -> Self {
+        let mut result = [0_u8; 32];
+        let id_bytes: [u8; 24] = asset.faucet_id.into();
+        result[..24].copy_from_slice(&id_bytes);
```

Comment:
Should we `debug_assert` here that the first bit is `0`? This is more of a `AccountId` problem, but might be nice to assert here too

Reply:
I added the same `Self::validate()` check here as for NFAs. It's a bit redundant - but I think that's OK.